src/server/game/Entities/Creature/Creature.cpp

```diff
@@ -2669,6 +2669,13 @@ CreatureMovementData const& Creature::GetMovementTemplate() const
     return GetCreatureTemplate()->Movement;
 }
 
+bool Creature::CanEnterWater() const
+{
```

Comment:
Isn't there CanSwim already? Why not just adding the pet check there instead?

Reply:
I don't know the coding involved, but there is a technical difference between creatures being blocked (by their AI) from entering water and the same creatures being able to swim if you spawn them in water.

Reply:
CanEnterWater() means a Unit can walk into water mesh, i.e. setting the NAV_WWATER filter in MMAPs to allow crossing rivers/lakes and reach enemies in water.

CanSwim() means the Unit is able to lift itself above sea bottom and reach targets floating in the water.

An Unit with CanEnterWater() but without CanSwim() will walk at the bottom of the sea, like a crab for example, and the Z coordinate returned by MMAPs will be GetGroundHeight() and not GetGroundOrWaterLevel().
In current TC crabs sometime get water level,  making them teleport at water level and then instantly sink to the bottom (as they are missing the flags that would allow them to swim)

Reply:
The 'teleport' is not properly solved by using the unit flag. It's actually handled by a currently wrong handled spline flag: Walkmode which is actually CanSwim. It has been corrected on master already and I've also backported it for 4.3.4 (and it solves the visual issue you describe) and I am sure this applies to 335 as well.

As for the ground underwater movement I would simply go with Groundlevel by default if CanSwim returns false or the unitflag is not set.

Reply:
> As for the ground movement, I would simply go with groundlevel if CanSwim returns false. No need to do additional fancy magic. If a unit is not able to swim it should stick on the ground in all cases. Just move the unit flag and pet check to the CanSwim helper and be done with it and keep the path generator changes to ignore water level of CanSwim returns false.

Sounds plain and simple. If no swim flag, it walks, no matter, there is water or not.

Reply:
This is already in 335 and that didn't fix the issue